src/core/SampleBuffer.cpp

```diff
@@ -957,12 +925,18 @@ void SampleBuffer::visualize( QPainter & _p, const QRect & _dr,
 	const int xb = _dr.x();
 	const int first = focus_on_range ? _from_frame : 0;
 	const int last = focus_on_range ? _to_frame : m_frames;
-	for( int frame = first; frame < last; frame += fpp )
-	{
-		l[n] = QPointF( xb + ( (frame - first) * double( w ) / nb_frames ),
-			( yb - ( m_data[frame][0] * y_space * m_amplification ) ) );
-		r[n] = QPointF( xb + ( (frame - first) * double( w ) / nb_frames ),
-			( yb - ( m_data[frame][1] * y_space * m_amplification ) ) );
+	
+	const bool isReversed = m_reversed;
+
+	auto startFrame = isReversed ? last : first;
+	auto count = (last - first) / fpp + 1;
+	auto step = isReversed ? -fpp : fpp;
+	for (int frame = startFrame, i = 0; i < count; frame += step, i++)
+	{
+		l[n] = QPointF (xb + ((frame - first) * double(w) / nb_frames),
+			(yb - (m_data[frame][0] * y_space * m_amplification)));
+		r[n] = QPointF (xb + ((frame - first) * double(w) / nb_frames),
+			(yb - (m_data[frame][1] * y_space * m_amplification)));
 		++n;
 	}
```

Comment:
This was already discussed on Discord, but since there were lots of messages where I was still figuring some things out I'm trying to make a more organized post here. One of the things that makes this section of the code confusing is that the naming of variables is very misleading. `_to_frame` for example isn't the last frame to be drawn, but instead points to the frame after the last frame to be drawn (like `std::end`, same applies to `m_frame` since it's the length of the buffer).

Another thing I noticed that completely went over our heads is that we didn't need to rewrite this loop at all. When the sample is reversed, the frames inside `m_data` are already reversed by using a call to `std::reverse`. So everything we had to do was draw them in the order they are laid in the buffer. The only reason the updated loop is actually drawing properly (and not drawing the reverse of an already reversed buffer) is the calculation of the X position of the point: Because we had `frame - first` it was actually not making any difference that we were drawing from the last to the first frame, because we were placing them in the same X position as before.

This section of the code doesn't need to be changed, we don't even need to check if the sample is reversed. But I'd propose the following changes:

```suggestion
	const int first = focus_on_range ? _from_frame : 0;
	const int last = focus_on_range ? _to_frame - 1 : m_frames - 1;

	for (int frame = first; frame <= last; frame += fpp)
	{
		auto x = xb + ((frame - first) * double(w) / nb_frames);
		// Partial Y calculation
		auto py = y_space * m_amplification;
		l[n] = QPointF(x, (yb - (m_data[frame][0] * py)));
		r[n] = QPointF(x, (yb - (m_data[frame][1] * py)));
		++n;
	}
```

The only changes to the loop we had originally is to make `last` point to the last frame instead of the frame after the last and @Veratil 's suggestion of moving some repeating calculation to outside the method call. I haven't tested it yet, but I believe it will draw reversed samples properly. Could you make the changes and test it @russiankumar ?

There are other changes I wish to do, like renaming `_to_frame` to something more accurate (or making it actually point to the last frame to be drawn), but that's outside the scope of this PR and would probably fit better in the refactor draft I've going locally.

Reply:
Alright, I'll be making these changes, and see if these work. I'll edit this comment if it does, and if it doesn't, I'll implement the two loops idea we were discussing on Discord.

Reply:
Awesome, I wouldn't worry too much about the name. I'm kind of doing a small refactor of `SampleBuffer.cpp` that I plan to submit after you merge this one and I fix the conflicts, I could change it there. Also, changing the name might span a few more files (the header and I think some AFP files).